url47fcfc0133_baliosian_tau_fst_tgz-pJ8-uy_edu_fing_mina_fsa_test_old_TestComposition1J8/src/uy/edu/fing/mina/fsa/tf/CompositeTf.java

```diff
@@ -0,0 +1,531 @@
+                newrr.left  = ((CompositeTf) ret.right).left; 
+                newrr.right = ((CompositeTf) ret.right).right; 
+                newrr.op    = Operator.AND;
+                
+                newl.left  = ((CompositeTf) ret.left).left;
+                newl.right = newlr;
+                newl.op    = Operator.AND;
+
+                newr.left  = ((CompositeTf) ret.left).right;
+                newr.right = newrr;
+                newr.op    = Operator.AND;
+                
+                ret.left  = newl;
+                ret.right = newr;
+                ret.op    = Operator.OR;
+                
+                return ret;
+              }
+            } else { //left is not OR 
+              if (((CompositeTf) ret.right).op.equals(Operator.OR)) {
+                //  (a∧b)∧(cvd)=((a∧b)∧c)v((a∧b)∧d) 
+                
+                CompositeTf newl  = new CompositeTf();
+                CompositeTf newr  = new CompositeTf();
+                
+                CompositeTf newll = new CompositeTf();
+                CompositeTf newrl = new CompositeTf();
+                
+                newll.left  = ((CompositeTf) ret.left).left; 
+                newll.right = ((CompositeTf) ret.left).right; 
+                newll.op    = Operator.AND;
+                
+                newrl.left  = ((CompositeTf) ret.left).left; 
+                newrl.right = ((CompositeTf) ret.left).right; 
+                newrl.op    = Operator.AND;
+                
+                newl.left  = newll;
+                newl.right = ((CompositeTf) ret.right).left;
+                newl.op    = Operator.AND;
+
+                newr.left  = newrl;
+                newr.right = ((CompositeTf) ret.right).right;
+                newr.op    = Operator.AND;
+                
+                ret.left  = newl;
+                ret.right = newr;
+                ret.op    = Operator.OR;
+                
+                return ret;
+              } else { // both are AND  
+                ret.op = op;
+                return ret; 
+              }
+            } 
+          } else { // right is simple
+            ret.right = (SimpleTf) right.clone();
+            if (((CompositeTf) ret.left).op.equals(Operator.OR)) {
+              // (b∨c)∧a=(a∧b)∨(a∧c)
+              
+              CompositeTf newltf = new CompositeTf();
+              newltf.left = ret.right;
+              newltf.right = ((CompositeTf) ret.left).left;
+              newltf.op = Operator.AND;
+              
+              CompositeTf newrtf = new CompositeTf();
+              newrtf.left = ret.right;
+              newrtf.right = ((CompositeTf) ret.left).right;
+              newrtf.op = Operator.AND;
+              
+              ret.left = newltf;
+              ret.right = newrtf;
+              ret.op = Operator.OR;
+              
+              return ret;
+
+            } else { // both are AND
+              ret.op = op;
+              return ret;
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 460:482, 488:510.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.